test/tests/functional/pbs_job_array.py

```diff
@@ -68,3 +69,51 @@ def test_arrayjob_Erecord_startval(self):
         # Verify that the value of 'start' isn't 0
         self.assertNotEqual(start_val, 0,
                             "E record value of 'start' for arrayjob is 0")
+
+    def test_arraysubjob_survive_restart(self):
+        """
+        Test to check if subjobs of an array job survive a pbs_server restart
+        """
+        a = {'resources_available.ncpus': 1}
+        self.server.manager(MGR_CMD_SET, NODE, a, self.mom.shortname)
+        j = Job(TEST_USER, attrs={
+            ATTR_J: '1-3', 'Resource_List.select': 'ncpus=1'})
+
+        j.set_sleep_time(5)
+
+        j_id = self.server.submit(j)
+
+        # 1. check job array has begun
+        self.server.expect(JOB, {'job_state': 'B'}, j_id)
+
+        # 2. wait till subjob 2 starts running
+        self.server.expect(JOB, {'job_state': 'R'}, create_subjob_id(j_id, 2))
+
+        # 3. Kill the server
+        self.server.stop('-KILL')
+
+        # 4. start the server
+        self.server.start()
+
+        # 5. array job should be B
+        self.server.expect(JOB, {'job_state': 'B'}, j_id, max_attempts=1)
```

Comment:
Having max_attempts=1 is very less. Assume the case where 100,000 jobs are getting retrieved from PBS server and also history is on. I suggest better don't use max_attempts here just depend on default value which is very large and suffices.

Reply:
Actually as per the use case, we have to have job_state as B in the first attempt itself. If it is not B in the first attempt then the test should be failed. Hence max_attempts=1 is mandatory. This will work properly irrelevant of number jobs in history or the load on the test machine.

Reply:
In case of slow running machine PTL framework fetching the job_state i.e. to do a selstat itself might take more time due to which I still feel that max_attempts=1 not enough.

Reply:
If it is suspected that selstat itself might take more time then we need to be concerned on time related parameter and not the max_attempts.

Reply:
In the interest of time what you want to do either you want to introduce time related parameter or use max_attempts to bigger value. I am fine with either of these approaches.

Reply:
As I already explained in previous comment that having max_attempts to bigger value defeats the purpose of the test case as we need to evaluate the job status in the first qstat call itself.  For time out, I feel the already existing PTL test timeout of 180 seconds should be fine.

Reply:
I think we should check for run_count here rather than depending on checking job_state in first attempt ?

Reply:
Actually for subjobs run_count do not increment, there is a bug open on it

Reply:
What is really required in PTL is to be able to detect a "timeout" error and allow the test writer to re-issue the call in case of timeout. This is different from the general expected result not matching. Anyways, for a local PTL test i do feel that the timeout is fine, and we can change if we have issues later.